Bundles UI tests

AWitcherONS opened the pull request:
### What is the context of this PR?
Bundles User Interface Journey from creation to Approval

**Testing Strategy**
User/Auth
Bundle:- volume, and contents 
The feature tests should be completed in a timely manner avoiding duplication where possible 

 role, function (Search(volume is significant), Create, Edit/insert Edit/remove, send to moderation, preview, approve)
 for each of these function what text can be seen,  what fields to edit(insert/delete), what actions (save etc) 
 what happens when mandatory fields fail validations on actions

    User role Bundle Life Cycle Happy Path 1 bundle Authorisation summary
                                 | Bundles Search | Dashboard Search | Create | Edit | Preview | Approve |
        Publishing Admin         | Can            |                  | Can    | Can  | Can     | Can     !
            As Creator of Bundle | N/A            | N/A              | N/A    | N/A  | N/A     | Cannot  |
            Not Creator          | N/A            | N/A              | N/A    | N/A  | N/A     | Can     |
        Publishing Officer       | Can            | N/A              | Can    | Can  | Can     | N/A     |
            As Creator of Bundle | N/A            | N/A              | N/A    | N/A  | N/A     | Cannot  |
            Not Creator          | N/A            | N/A              | N/A    | N/A  | N/A     | Can     |
        Viewer                   | N/A            | N/A              | Cannot | N/A  | N/A     | N/A     |
            not in preview team  | Cannot         | N/A              | N/A    | N/A  | Cannot  | N/A     |
            in preview team      | Can            | N/A              | N/A    | N/A  | Can     | N/A     |
/

### How to review
The Python code should comply with current standards
The feature tests run without failures and do not impact on run time


### Follow-up Actions
A Viewer cannot use the Search in Bundles menu there isn't an error but no know bundles are not found 
review on approval Authorisation of bundles

AWitcherONS committed:
add bundles UI

AWitcherONS committed:
add background

AWitcherONS committed:
resolve merge main conflicts

AWitcherONS committed:
create bundle

AWitcherONS committed:
add multiple pages for bundles index

AWitcherONS committed:
Merge branch 'main' into feature/bundles_UI_test-CMS-333

AWitcherONS committed:
commit before merge

AWitcherONS committed:
Merge branch 'main' into feature/bundles_UI_test-CMS-333

AWitcherONS committed:
another commit

AWitcherONS committed:
complete steps

AWitcherONS committed:
Merge branch 'main' into feature/bundles_UI_test-CMS-333

AWitcherONS committed:
resolve lint errors

MaciekBaron commented on functional_tests/features/bundle_ui.feature:
```diff
@@ -0,0 +1,267 @@
+        When the Publishing Admin logs in
+        And the user can see the Bundles menu item
+        And the user clicks the Bundles menu item
+        Then the user can search for unknown bundle with response "There are no bundles to"
```
Should this also say "There are no bundles to display" like on line 39? If it can say the same thing, then we don't need to parameterise our step as well.

MaciekBaron commented on functional_tests/steps/bundles.py:
```diff
@@ -121,3 +126,166 @@ def the_user_can_see_the_bundle_details_with_creator(context: Context) -> None:
+@step('the user can search for unknown bundle with response "{response}"')
+def search_for_new_bundle(context: Context, response: str) -> None:
+    unknown_bundle = "Unknown bundle"
+    if unknown_bundle not in get_bundles_names(context.bundles):
```
I don't think we should have an `if` statement like this here. This step will pass if the "Unknown bundle" is amongst the bundles in the context, without really doing what it is supposed to (i.e. check if the user can search and can see the adequate response).

MaciekBaron commented on functional_tests/steps/bundles.py:
```diff
@@ -121,3 +126,166 @@ def the_user_can_see_the_bundle_details_with_creator(context: Context) -> None:
+@given("there are {no_bundles} bundles")
+def multiple_bundles_create(context: Context, no_bundles: str) -> None:
+    context.bundles = []
+    if no_bundles.isdigit():
```
The step will fail silently if `no_bundles` is not a digit.

I think it's worth failing if it's not:
```python
if not no_bundles.isdigit():
    raise ValueError(f"Invalid parameter: '{no_bundles}' must be a digit.")
```

Same problem on line 143, 155 and 170.

MaciekBaron commented on functional_tests/steps/bundles.py:
```diff
@@ -121,3 +126,166 @@ def the_user_can_see_the_bundle_details_with_creator(context: Context) -> None:
+        for preview_team_index in range(int(no_preview_teams)):
+            context.teams.append(
+                Team.objects.create(
+                    identifier="preview-team-" + str(preview_team_index), name="Preview_Team_" + str(preview_team_index)
```
f-strings will be better here for readability:
```suggestion
                    identifier=f"preview-team-{preview_team_index}", name=f"Preview_Team_{preview_team_index}"
```

We could also perhaps drop the underscores?

```suggestion
                    identifier=f"preview-team-{preview_team_index}", name=f"Preview Team {preview_team_index}"
```

MaciekBaron commented on functional_tests/steps/bundles.py:
```diff
@@ -121,3 +126,166 @@ def the_user_can_see_the_bundle_details_with_creator(context: Context) -> None:
+            context.release_calendar_pages.append(
+                ReleaseCalendarPageFactory(
+                    release_date=nowish,
+                    title="Release Calendar Page" + str(release_calendar_index),
```
Similar to above, f-string a better (and I think there's a missing space):

```suggestion
                    title=f"Release Calendar Page {release_calendar_index}",
```

Same for line 174.

MaciekBaron commented on functional_tests/steps/bundles.py:
```diff
@@ -121,3 +126,166 @@ def the_user_can_see_the_bundle_details_with_creator(context: Context) -> None:
+    context.user_data = create_user(user_role)
+
+
+@given("the {user_role} is a member of the Preview teams")
```
You've defined a parameter here `{user_role}` but it doesn't seem to be used by the function itself, so I think it can be dropped. If we keep it as "the user is a member of the Preview teams", it should be clear from the context that it's the user we created.

MaciekBaron commented on functional_tests/steps/bundles.py:
```diff
@@ -121,3 +126,166 @@ def the_user_can_see_the_bundle_details_with_creator(context: Context) -> None:
+
+@given("the {user_role} is a member of the Preview teams")
+def add_user_to_preview_teams(context: Context) -> None:
+    for team in context.teams:
+        user = context.user_data["user"]
+        user.teams.add(team)
```
This can be simplified:
```suggestion
    user = context.user_data["user"]
    user.teams.add(*context.teams)
```
`.add()` allows us to provide multiple objects for optimisation.

https://docs.djangoproject.com/en/5.2/ref/models/relations/#django.db.models.fields.related.RelatedManager.add

MaciekBaron commented on functional_tests/steps/bundles.py:
```diff
@@ -121,3 +126,166 @@ def the_user_can_see_the_bundle_details_with_creator(context: Context) -> None:
+    context.page.get_by_role("textbox", name="Search term").click()
+    context.page.get_by_role("textbox", name="Search term").fill(context.teams[1])
+    context.page.get_by_text(context.teams[1]).click()
+    context.page.get_by_role("button", name="Confirm selection").click()
```
I think we should follow this with an assertion to confirm that we have added the preview team successfully.

AWitcherONS committed:
pre_refinement

AWitcherONS committed:
Merge branch 'main' into feature/bundles_UI_test-CMS-333

AWitcherONS committed:
refactor 1

AWitcherONS committed:
Merge branch 'main' into feature/bundles_UI_test-CMS-333

AWitcherONS committed:
refactor 2

AWitcherONS committed:
refactor 3

AWitcherONS committed:
Merge branch 'main' into feature/bundles_UI_test-CMS-333

AWitcherONS committed:
Merge branch 'main' into feature/bundles_UI_test-CMS-333

AWitcherONS committed:
review resolution

AWitcherONS committed:
Merge branch 'main' into feature/bundles_UI_test-CMS-333

AWitcherONS committed:
some lint resolutions

AWitcherONS committed:
some more lint resolutions

AWitcherONS committed:
even some more lint resolutions

AWitcherONS committed:
format  resolutions

AWitcherONS committed:
Merge branch 'main' into feature/bundles_UI_test-CMS-333

AWitcherONS committed:
lint resolution v3

AWitcherONS committed:
Merge branch 'main' into feature/bundles_UI_test-CMS-333

AWitcherONS committed:
Merge branch 'main' into feature/bundles_UI_test-CMS-333

AWitcherONS committed:
modification of json struct

AWitcherONS committed:
lint changes

AWitcherONS committed:
resolve test failures 1

AWitcherONS committed:
resolve test failures 2

AWitcherONS committed:
resolve test failures 3

AWitcherONS committed:
resolve test failures 4

AWitcherONS committed:
resolve test failures 5

AWitcherONS committed:
interim commit

AWitcherONS committed:
Merge branch 'main' into feature/bundles_UI_test-CMS-333

AWitcherONS committed:
Pre Time out commit

AWitcherONS committed:
Pre Time out commit

AWitcherONS committed:
Pre Time out commit

AWitcherONS committed:
Merge branch 'main' into feature/bundles_UI_test-CMS-333

AWitcherONS committed:
Pre Time out commit

AWitcherONS committed:
Pre Time out commit

AWitcherONS committed:
functional tests with failures in edit and preview

AWitcherONS committed:
functional tests with failures in edit and preview

AWitcherONS committed:
functional tests with failures in edit and preview

MaciekBaron commented on functional_tests/steps/bundles.py:
```diff
@@ -121,3 +128,267 @@ def the_user_can_see_the_bundle_details_with_creator(context: Context) -> None:
+@given("there are {number_of_bundles} bundles with {bundle_details}")
+def multiple_bundles_create(context: Context, number_of_bundles: str, bundle_details: str) -> None:
+    bundle_dets = {}
+    if json_str_to_dict(bundle_details):
+        bundle_dets = json.loads(bundle_details)
```
This looks a bit strange to me, the function you call does `json.loads` and then you catch the exception which returns false. You then attempt json loads again if false is not returned.

I don't think this approach works well. From my point of view this function would never receive invalid json, unless there's a mistake.

MaciekBaron commented on functional_tests/features/bundle_ui.feature:
```diff
@@ -0,0 +1,130 @@
+           | number_of_bundles | Role                       | Creator Role        | Bundle_Details                                                                                                                                              |
+           | 1                 | Publishing Admin           | Publishing Admin    | {"Role": "Publishing Admin",   "Creator Role": "Publishing Admin",   "status": "In_Review", "preview_teams": true,  "add_rel_cal": true, "add_stat_page": true} |
+           | 1                 | Publishing Officer         | Publishing Admin    | {"Role": "Publishing Officer", "Creator Role": "Publishing Admin",   "status": "In_Review", "preview_teams": true,  "add_rel_cal": true, "add_stat_page": true} |
+#           | 1                 | Viewer                     | Publishing Admin    | {"Role": "Publishing Admin",   "Creator Role": "Publishing Admin",   "status": "In_Review", "preview_teams": true,  "add_rel_cal": true, "add_stat_page": true} |
```
You've marked the PR as ready for review but there's still commented out code here.

MaciekBaron replied:
Sorry, where's the reason posted?

MaciekBaron replied:
This information should really be in the PR description so that reviewers know what's up. If you could add it there, and then add a TODO, that would be grand, thanks.

MaciekBaron commented on functional_tests/steps/bundles.py:
```diff
@@ -121,3 +128,267 @@ def the_user_can_see_the_bundle_details_with_creator(context: Context) -> None:
+def multiple_statistical_analysis(context: Context, no_statistical_analysis: str) -> None:
+    context.statistical_article_pages = []
+    context.article_series_page = ArticleSeriesPageFactory(title="PSF")
+    if no_statistical_analysis.isdigit():
```
Do we need to check if `no_statistical_analysis` is a digit? It seems to me that it should always be a digit.

MaciekBaron replied:
Could you show me in what scenario that would not be an integer?

AWitcherONS replied:
There is not a scenario where this would not be an integer just instinct where input is so precarious the same with json.loads

MaciekBaron commented on functional_tests/steps/bundles.py:
```diff
@@ -121,3 +128,267 @@ def the_user_can_see_the_bundle_details_with_creator(context: Context) -> None:
+    context.page.get_by_title("View").click()
+
+    expect(context.page.get_by_text("Name")).to_be_visible()
+    # expect(context.page.get_by_role("term").filter(has_text=re.compile(r"^Status$"))).to_be_visible()
```
More commented out code here

MaciekBaron commented on functional_tests/steps/bundles.py:
```diff
@@ -121,3 +128,267 @@ def the_user_can_see_the_bundle_details_with_creator(context: Context) -> None:
+    expect(context.page.locator("#latest-bundles-content")).to_contain_text(context.bundles[0].name)
+    context.page.get_by_title("Edit this bundle").click()
+    context.page.get_by_role("button", name="More actions").click()
+    context.page.locator("#panel-status-content").click()
```
I've already sent a comment to you about this. It is strange to perform actions at the end of a function which is designed to make assertions. Please revisit this.

MaciekBaron replied:
Hi Ann, just to clarify what I mean here: your assertion functions shouldn't end with actions but rather assertions using `expect()` - you should define what you expect to see rather than try to click on something and see if it works. For instance, check if the "Edit this bundle" link is visible and active etc.. Clicking on it doesn't tell us much.

MaciekBaron commented on functional_tests/steps/bundles.py:
```diff
@@ -121,3 +128,267 @@ def the_user_can_see_the_bundle_details_with_creator(context: Context) -> None:
+def add_article_page_in_edit(context: Context) -> None:
+    # add Article
+    if hasattr(context, "statistical_article_pages"):
+        expect(context.page.get_by_role("heading", name="Bundled pages").locator("span")).to_be_visible()
```
The `expect` calls here are redundant, you are performing an action, not making a check

MaciekBaron replied:
I think you mean "unit tests" instead of "functional tests"? As these are functional tests.
These `expect` calls are not needed because your actions will fail if the element doesn't exist, so the calls are redundant. If you look at other code you'll see we don't follow this pattern.

MaciekBaron commented on functional_tests/steps/bundles.py:
```diff
@@ -121,3 +128,267 @@ def the_user_can_see_the_bundle_details_with_creator(context: Context) -> None:
+        expect(context.page.get_by_role("checkbox", name=context.teams[0].name)).to_be_visible()
+        context.page.get_by_role("checkbox", name=context.teams[0].name).check()
+        context.page.wait_for_timeout(100)
+        expect(context.page.get_by_role("button", name="Confirm selection")).to_be_visible()
```
You don't need to call `expect` before clicking a button

AWitcherONS committed:
remove timeout

AWitcherONS committed:
resolve pr comments 1

nehakerung commented on functional_tests/steps/bundles.py:
```diff
@@ -121,3 +128,272 @@ def the_user_can_see_the_bundle_details_with_creator(context: Context) -> None:
+            BundleTeam.objects.create(parent=bundle, team=team)
+
+
+def json_str_to_dict(bundle_details):
```
Would it be better to have the helper functions in their own step_helpers file

MaciekBaron replied:
I would remove this function all together

nehakerung commented on functional_tests/steps/bundles.py:
```diff
@@ -121,3 +128,272 @@ def the_user_can_see_the_bundle_details_with_creator(context: Context) -> None:
+def the_user_cannot_add_bundles(context: Context) -> None:
+    expect(context.page.get_by_role("link", name="Add bundle")).not_to_be_visible()
+    expect(context.page.get_by_text("Bundles ready for preview")).to_be_visible()
+    expect(context.page.get_by_text("There are currently no")).to_be_visible()
```
Could check for the full error message text

nehakerung commented on functional_tests/steps/bundles.py:
```diff
@@ -121,3 +128,272 @@ def the_user_can_see_the_bundle_details_with_creator(context: Context) -> None:
+        context.page.get_by_role("button", name="Approve").click()
+
+
+def create_user_by_role(context, user_role):
```
Missing type hints(!)

nehakerung commented on functional_tests/steps/bundles.py:
```diff
@@ -121,3 +128,272 @@ def the_user_can_see_the_bundle_details_with_creator(context: Context) -> None:
+    expect(context.page.get_by_text("Datasets", exact=True)).to_be_visible()
+
+
+@step("the user cannot preview a bundle")
+def cannot_preview_bundle(context: Context) -> None:
+    context.page.get_by_role("link", name="Bundles", exact=True).click()
+    expect(context.page.get_by_role("button", name=f"More options for '{context.bundles[0].name}'")).not_to_be_visible()
+
+
+@step("the user cannot approve a bundle")
+def cannot_approve_bundle(context: Context) -> None:
+    context.page.get_by_role("link", name="Bundles", exact=True).click()
+    expect(context.page.get_by_role("button", name=f"More options for '{context.bundles[0].name}'")).not_to_be_visible()
```
These functions are the same-  if thats correct you can add the steps like so
```
@step("the user cannot preview a bundle")
@step("the user cannot approve a bundle")
def ...
```

MaciekBaron commented on functional_tests/steps/bundles.py:
```diff
@@ -121,3 +128,272 @@ def the_user_can_see_the_bundle_details_with_creator(context: Context) -> None:
+        context.page.get_by_role("button", name="Confirm selection").click()
+
+
+def add_reslease_calendar_in_edit(context: Context) -> None:
```
Typo
```suggestion
def add_release_calendar_in_edit(context: Context) -> None:
```

MaciekBaron commented on functional_tests/steps/bundles.py:
```diff
@@ -121,3 +128,272 @@ def the_user_can_see_the_bundle_details_with_creator(context: Context) -> None:
+            BundlePage.objects.create(parent=bundle, page=page)
+
+
+def add_release_calendar(context: Context):
```
Missing return type
```suggestion
def add_release_calendar(context: Context) -> None:
```

MaciekBaron commented on functional_tests/steps/bundles.py:
```diff
@@ -121,3 +128,272 @@ def the_user_can_see_the_bundle_details_with_creator(context: Context) -> None:
+        expect(context.page.locator("#latest-bundles-heading")).to_contain_text("Latest active bundles")
+        expect(context.page.locator("#latest-bundles-content")).to_contain_text(context.bundles[0].name)
+        context.page.get_by_title("Edit this bundle").click()
+        context.page.get_by_role("button", name="More actions").click()
+        context.page.get_by_role("button", name="Approve").click()
```
You can take these two lines out of the if/else block so that you don't repeat yourself

MaciekBaron commented on functional_tests/steps/bundles.py:
```diff
@@ -121,3 +128,272 @@ def the_user_can_see_the_bundle_details_with_creator(context: Context) -> None:
+
+
+@given("there are {number_of_bundles} bundles with {bundle_details}")
+def multiple_bundles_create(context: Context, number_of_bundles: str, bundle_details: str) -> None:
+    bundle_dets = json_str_to_dict(bundle_details)
+    if bundle_dets:
+        context.bundles = []
+
+        if number_of_bundles.isdigit():
+            for __ in range(int(number_of_bundles)):
+                if not next(item for item in context.users if bundle_dets["Creator Role"] in item):
+                    context.users.append({bundle_dets["Creator Role"]: create_user(bundle_dets["Creator Role"])})
+
+                bundle_creator = next(item for item in context.users if bundle_dets["Creator Role"] in item)
+                bundle_status = BundleStatus.DRAFT
+                bundle_approved = False
+                if bundle_dets["status"] == "Approved":
+                    bundle_status = BundleStatus.APPROVED
+                    bundle_approved = True
+                if bundle_dets["status"] == "In_Review":
+                    bundle_status = BundleStatus.IN_REVIEW
+                bundle = BundleFactory(
+                    created_by=bundle_creator.get("user"),
+                    status=bundle_status,
+                    approved=bundle_approved,
+                )
+
+                if bool(bundle_dets["preview_teams"]) and hasattr(context, "teams"):
+                    add_teams(context)
+
+                if bool(bundle_dets["add_rel_cal"]) and hasattr(context, "release_calendar_pages"):
+                    add_release_calendar(context)
+
+                if bool(bundle_dets["add_stat_page"]) and hasattr(context, "statistical_article_pages"):
+                    add_article_pages(context)
+
+                context.bundles.append(bundle)
```
I've been reading this code and trying to understand it.

I see on line 194 you enter a for loop for the range `number_of_bundles`.

Within the for loop, you create the bundle on line 206, and append it to `context.bundles` on line 221.

However the things which happen around that don't seem to belong to the for loop:

- Lines 195-196 you create a creator role if one doesn't exist, this should be outside of the loop
- Lines 212-213 you call `add_teams` which creates a `BundleTeam` for each team and bundle, but since you are running this in a loop you recall this function on every iteration, recreating `BundleTeam` entries for existing bundles and teams
- Ditto on lines 215-216, you create unnecessary `BundlePage`s
- Ditto on lines 218-219

It seems more logical to me to take this code out of the loop so it only runs once for all bundles and pages, since the functions you call (`add_teams` etc.) already loop through your objects.

MaciekBaron reviewed (changes requested):
Could you also flesh out the PR description and include that comment you made on Slack about some of the scenarios failing and still being work in progress (you marked them as TODO for now)

AWitcherONS committed:
Merge branch 'main' into feature/bundles_UI_test-CMS-333

AWitcherONS committed:
Merge branch 'main' into feature/bundles_UI_test-CMS-333

AWitcherONS committed:
resolve comments 3

AWitcherONS committed:
Merge branch 'main' into feature/bundles_UI_test-CMS-333

AWitcherONS committed:
resolve pr comments 2

AWitcherONS committed:
resolve pr comments 2

AdamHawtin commented on functional_tests/features/bundle_ui.feature:
```diff
@@ -0,0 +1,133 @@
+    User Role Bundle Life Cycle Happy Path
+                                 | Search | Create | Edit | Preview | Approve |
+        Publishing Admin         | ToDo   | Can    | Can  | Can     | N/A     |
+            As Creator of Bundle | N/A    | N/A    | N/A  | N/A     | Cannot  |
```
To clarify something, currently the user who created the bundle _can_ approve the bundle themselves, it is pages that require a different user approval.

AdamHawtin reviewed (commented):
A lot of the test cases in the feature file seem to be commented out still, what is happening with these?

MaciekBaron commented:
> A lot of the test cases in the feature file seem to be commented out still, what is happening with these?

@AdamHawtin here's some additional context: https://github.com/ONSdigital/dis-wagtail/pull/273#discussion_r2307495188

MaciekBaron commented on functional_tests/steps/bundles.py:
```diff
@@ -121,3 +128,268 @@ def the_user_can_see_the_bundle_details_with_creator(context: Context) -> None:
+    context.page.get_by_role("button", name="Save to preview").click()
+
+
+def add_preview_team_in_edit(context: Context) -> None:
```
I think it might be worth moving these helper functions to a separate file in `step_helpers` just to follow our existing convention.

sanjeevz3009 commented on functional_tests/features/bundle_ui.feature:
```diff
@@ -0,0 +1,133 @@
+        Then the user can see the Bundles ready for preview
+        Then the <Role> can preview a bundle
+
+      Examples: bundles
```
am confused of what's going on here? And if this is needed?

As I been debugging line by line the behave steps, and using the playwight inspector to check for assertions etc, and so far, this is not making sense. Assertions seems to be confusing and some assertions are incomplete as it's checking for half of a message or text on a page etc.

sanjeevz3009 commented on functional_tests/features/bundle_ui.feature:
```diff
@@ -0,0 +1,133 @@
+           | 1                 | Viewer                     | Publishing Officer  | {"Role": "Viewer",             "Creator Role": "Publishing Officer", "status": "In_Review", "preview_teams": true,  "add_rel_cal": true, "add_stat_page": true} |
+
+
+      Scenario Outline: A User cannot preview a bundle
```
Why is there a nested scenario in here?

sanjeevz3009 replied:
Also similar to some other scenarios above, it doesn't seem to be doing anything?

sanjeevz3009 replied:
I tried to run this specific scenario and it won't run because it's nested under another scenario. So this scenario is basically not running. Even if it does, it will not be doing anything.

https://cucumber.io/docs
https://behave.readthedocs.io/en/latest/

sanjeevz3009 commented on functional_tests/features/bundle_ui.feature:
```diff
@@ -0,0 +1,133 @@
+
+#----- Bundle Approve UI Tests -----
+
+      Scenario Outline: A user can approve a bundle
```
This is not currently doing anything

sanjeevz3009 commented on functional_tests/features/bundle_ui.feature:
```diff
@@ -0,0 +1,133 @@
+#            | 1                 | Publishing Officer         | Publishing Admin    | {"Role": "Publishing Officer", "Creator Role": "Publishing Admin",   "status": "In_Review", "preview_teams": true, "add_rel_cal": true, "add_stat_page": true} |
+
+
+      Scenario Outline: A user cannot approve a bundle
```
Not doing anything

sanjeevz3009 reviewed (changes requested):
@MebinAbraham @MaciekBaron @AdamHawtin @AWitcherONS 
Did an initial round of review. This PR is far from complete. This doesn't really align with the bundle UI card, requirements and what we want out of it, the ticket.

Considering the bundles app in itself is a very custom feature developed by TBX, we will need to write really good scenarios and make sure our design of the step files, scenarios and the whole test foundation for bundles UI tests are really well laid out.

We would need to design the step definitions and the scenarios really well so that it's maintainable, readable, robust and future proof. You can take look at other UI tests, particularly RC UI tests where, you can build it out like lego blocks, where it very precise in what it tests, we can expand, such as add any big features to the RC page or remove or any small changes from adding new validation error messages on the RC page and we can reflect that in our scenarios/ test cases with ease, without increasing the file size, introducing complexity and going on a unmaintable path. These tests shouldn't be written in a sequential brute-force way; they should be as DRY as possible (some tests, such as for a small page or functionality, it's a different case). Bundles should have less complexity than the RC page, as that RC page was more complex to design in terms of the features it had, and has been tested thoroughly.

Otherwise, we take a huge risk of:
- Failing to catch any crucial regressions
- Bugs that may arise
- Any regressions caused by other features/apps in the codebase as bundle links heavily to so many other pages, services, and apps
  - RC pages and other pages you can add to the bundle
  - Datasets
  - Notifications
  - It all goes through a workflow
  - A couple of other things
  
If anything regresses, changes happen intentionally or unintentionally, then it can have a cascading effect on the bundle's functionality and how it behaves.
  
This is one of our key Django apps we have which is tied to scheduling publications, releases, and 7 am releases, and is one of the biggest goals of our product right now.

I would suggest taking a look at the release calendar page UI tests to see how we designed the scenarios, step definitions, step helpers, etc.

On some pages, such as the release calendar page, statistical article, methodology pages, bundles app/ page, we cannot take risks by not writing good quality tests/ inadequate tests that are not designed well and future-proof.

At least, even if we want to test something basic, writing the tests sequentially/ brute force (which will not be maintainable in the long run). I would expect something basic/ simple:
The below should be user-driven (driven by user interactions, unless for specific cases you can use factories)
- A user visits the bundles page (PO, PA, Viewer)
- If it's a user with the correct permissions -> they can schedule a bundle with the RC page (create the RC page via factories, we already have one, so you can just use that/ instantitate wherever necessary as we aren't testing RC page features here, so we don't care about that) -> add some pages (statsistical article, info, methodologies), add datasets etc -> we can take the bundle through various statuses successfully -> publish it etc (there are different ways achieve this)
-> Check if the bundle can be seen by a viewer in a particular preview team, inspect the bundle, and view the pages in the preview
-> Check for some basic validations, such as pages in the bundle have to be ready to publish/ approved through page workflows -> if not, we shouldn't be able to schedule a bundle
-> In a nutshell, checking the bundles pages/ apps functionality from a user's perspective in a comprehensive manner

I would suggest understanding, reviewing the bundle's features, what we can do, what we can't do and writing some basic scenarios to start with (the requirement of the bundle's UI ticket is to make sure we test this thoroughly).

Also, the above examples are only to give an idea, but there are way more test cases/ scenarios you will need to design.

Then you will have to make sure to design the code so that the scenarios, step definitions, and step helpers are robust, maintainable and future-proof.

AWitcherONS committed:
Merge branch 'main' into feature/bundles_UI_test-CMS-333

MaciekBaron commented on functional_tests/steps/bundles.py:
```diff
@@ -121,3 +128,268 @@ def the_user_can_see_the_bundle_details_with_creator(context: Context) -> None:
+    if hasattr(context, "search_release_calendar"):
+        expect(context.page.get_by_role("button", name="Choose Release Calendar page")).to_be_visible()
+        context.page.get_by_role("button", name="Choose Release Calendar page").click()
+        context.page.get_by_role("textbox", name="Search term").click()
```
We don't need to `click()` on elements to `fill()` them (here and elsewhere).

AWitcherONS committed:
resolve PR comments

AWitcherONS committed:
Merge branch 'main' into feature/bundles_UI_test-CMS-333

AWitcherONS committed:
restore poetry.lock

AWitcherONS committed:
add tests from cms-333

AWitcherONS committed:
resolve comments

AWitcherONS committed:
resolve comments

AWitcherONS committed:
Merge branch 'main' into feature/bundles_UI_test-CMS-333

AWitcherONS committed:
Merge branch 'main' into feature/bundles_UI_test-CMS-333

AWitcherONS committed:
comment resolutions

AWitcherONS committed:
Merge branch 'main' into feature/bundles_UI_test-CMS-333

AWitcherONS commented on functional_tests/steps/bundles.py:
```diff
@@ -121,3 +128,268 @@ def the_user_can_see_the_bundle_details_with_creator(context: Context) -> None:
+                    approved=bundle_approved,
+                )
+
+                if bool(bundle_dets["preview_teams"]) and hasattr(context, "team"):
```
removed the json function
Thanks

AWitcherONS commented on functional_tests/steps/bundles.py:
```diff
@@ -121,3 +128,268 @@ def the_user_can_see_the_bundle_details_with_creator(context: Context) -> None:
+    expect(context.page.get_by_text("There are currently no")).to_be_visible()
+
+
+@step("the user can find the bundle")
```
Done

AWitcherONS commented on functional_tests/steps/bundles.py:
```diff
@@ -121,3 +128,268 @@ def the_user_can_see_the_bundle_details_with_creator(context: Context) -> None:
+        context.page.get_by_role("button", name="Add preview team").click()
+        expect(context.page.get_by_role("textbox", name="Search term")).to_be_visible()
+        context.page.get_by_role("textbox", name="Search term").click()
+        context.page.get_by_role("textbox", name="Search term").fill(context.teams[0].name)
```
There is only 1 team now and users are added or not by the scenario lines 
```
        And there is a preview team
        And the <role> is a member of the preview team
```
hope that has answered your confusion it did mine

AWitcherONS commented on functional_tests/steps/bundles.py:
```diff
@@ -121,3 +128,268 @@ def the_user_can_see_the_bundle_details_with_creator(context: Context) -> None:
+    expect(context.page.locator("#bundles-for-preview-heading")).to_contain_text("Bundles ready for preview")
+    expect(context.page.locator("#bundles-for-preview-content")).to_contain_text(context.bundles[0].name)
+    context.page.get_by_title("View").click()
+    context.page.wait_for_timeout(200)
```
I think this was for the dashboard in which playwright was having difficulties with a bundle in latest and ready for review so the pseudo search has had to be removed

AWitcherONS commented on functional_tests/steps/bundles.py:
```diff
@@ -121,3 +128,268 @@ def the_user_can_see_the_bundle_details_with_creator(context: Context) -> None:
+
+
+@given("there is a preview team")
+def multiple_preview_teams_create(context: Context) -> None:
```
See Above

MaciekBaron commented on functional_tests/steps/bundles.py:
```diff
@@ -121,3 +129,337 @@ def the_user_can_see_the_bundle_details_with_creator(context: Context) -> None:
+    context.page.get_by_role("link", name="Add bundle").click()
+    context.page.get_by_role("button", name="Save as draft").click()
+    expect(context.page.get_by_role("status")).to_contain_text("The bundle could not be created due to errors.")
+    expect(context.page.locator("#panel-child-name-errors")).to_contain_text("This field is required.")
```
Ideally we would split this to two steps
* the logged in user saves a bundle
* the logged in user sees a validation error

This allows you to avoid repetition, and better structures your test with steps that do something and assertions that check something.

MaciekBaron commented on functional_tests/steps/bundles.py:
```diff
@@ -121,3 +129,337 @@ def the_user_can_see_the_bundle_details_with_creator(context: Context) -> None:
+@given("there is a Statistical Analysis page")
+def statistical_analysis(
+    context: Context,
+) -> None:
```
This can be on one line.

MaciekBaron commented on functional_tests/steps/bundles.py:
```diff
@@ -121,3 +129,337 @@ def the_user_can_see_the_bundle_details_with_creator(context: Context) -> None:
+            bundle.save()
+            context.bundles.append(bundle)
+
+        sorted(context.bundles, key=lambda x: x.name)
```
The `sorted()` function _returns_ a sorted list.

https://docs.python.org/3/library/functions.html#sorted

MaciekBaron commented on functional_tests/steps/bundles.py:
```diff
@@ -121,3 +129,337 @@ def the_user_can_see_the_bundle_details_with_creator(context: Context) -> None:
+            elif bundle_dets["status"] == "In_Review":
+                bundle_status = BundleStatus.IN_REVIEW
+
+            if bundle_dets["creator_role"] != "":
```
Can `bundle_dets["creator_role"]` ever be an empty string?

MaciekBaron commented on functional_tests/steps/bundles.py:
```diff
@@ -121,3 +129,337 @@ def the_user_can_see_the_bundle_details_with_creator(context: Context) -> None:
+                bundle.approved = bundle_approved
+            bundle.save()
+
+            if bool(bundle_dets["preview_teams"]) and hasattr(context, "team"):
```
Why do you check `hasattr(context, "team")`? Isn't that guaranteed by `there is a preview team` which is called in every scenario using this step?

MaciekBaron commented on functional_tests/steps/bundles.py:
```diff
@@ -121,3 +129,337 @@ def the_user_can_see_the_bundle_details_with_creator(context: Context) -> None:
+                context.bundleteams.append({"parent": bundle, "team": context.team})
+            bundle.save()
+
+            if bool(bundle_dets["add_rel_cal"]) and hasattr(context, "release_calendar_page"):
```
If your bundle details have `"add_rel_cal"` as true, but you forgot to add a step to add a release calendar, this indicates a bug in the test code. Don't check for `hasattr(context, "release_calendar_page")`, fail here so that you can see your mistake.

MaciekBaron commented on functional_tests/steps/bundles.py:
```diff
@@ -121,3 +129,337 @@ def the_user_can_see_the_bundle_details_with_creator(context: Context) -> None:
+    if bundle_dets:
+        context.bundles = []
+        context.bundlepages = []
+        context.bundleteams = []
```
For consistency can't we have:
```suggestion
        context.bundle_pages = []
        context.bundle_teams = []
```

MaciekBaron replied:
Second point here, why save this in `context` if you don't use it anywhere else? Am I missing something?

AWitcherONS committed:
Merge branch 'main' into feature/bundles_UI_test-CMS-333

AWitcherONS commented on functional_tests/features/bundle_ui.feature:
```diff
@@ -0,0 +1,133 @@
+#           | Publishing Officer |
+#
+#
+    Scenario Outline: A User cannot create a bundle
```
Please see above

AWitcherONS committed:
correct approval steps as per Pr comment

AWitcherONS committed:
merge main

MaciekBaron commented on functional_tests/steps/bundles.py:
```diff
@@ -10,14 +11,15 @@
-from cms.bundles.tests.factories import BundleFactory, BundlePageFactory
-from cms.release_calendar.enums import ReleaseStatus
+from cms.bundles.models import BundleTeam
+from cms.bundles.tests.factories import BundleFactory
```
This looks like a bad merge. You removed `BundlePageFactory`, `ReleaseStatus` etc. even though those are used.

This would have been [picked up by the linter](https://github.com/ONSdigital/dis-wagtail/actions/runs/19671912859/job/56342699989?pr=273). Please remember to run linters before pushing code.

AWitcherONS committed:
merge main

AWitcherONS committed:
merge main

AWitcherONS committed:
Merge branch 'main' into feature/bundles_UI_test-CMS-333

AWitcherONS committed:
merge fix

AWitcherONS committed:
merge fix

AWitcherONS committed:
merge fix

AWitcherONS committed:
merge fix

AWitcherONS committed:
merge fix

AWitcherONS committed:
merge fix

AWitcherONS committed:
main merge

AWitcherONS committed:
Merge branch 'main' into feature/bundles_UI_test-CMS-333

AWitcherONS committed:
approval push

AWitcherONS committed:
correct lint failure

AWitcherONS committed:
lint resolutions

MaciekBaron commented on functional_tests/features/bundle_ui.feature:
```diff
@@ -0,0 +1,211 @@
+#        | 1                 | Publishing Admin           | Publishing Officer  | {"role": "Publishing Admin",   "creator_role": "Publishing Officer", "status": "In_Review", "preview_teams": true, "add_rel_cal": true, "add_stat_page": true} |
+#        | 1                 | Publishing Admin           | Publishing Admin    | {"role": "Publishing Admin",   "creator_role": "Publishing Admin",   "status": "In_Review", "preview_teams": true, "add_rel_cal": true, "add_stat_page": true} |
+#        | 1                 | Publishing Officer         | Publishing Officer  | {"role": "Publishing Officer", "creator_role": "Publishing Officer", "status": "In_Review", "preview_teams": true, "add_rel_cal": true, "add_stat_page": true} |
+#        | 1                 | Publishing Officer         | Publishing Admin    | {"role": "Publishing Officer", "creator_role": "Publishing Admin",   "status": "In_Review", "preview_teams": true, "add_rel_cal": true, "add_stat_page": true} |
```
If there is a reason why you have this commented out, it would be prudent to either leave a comment explaining why, or update the PR description so that this is clear.

Same on line 62, 63 and 211.

MaciekBaron commented on functional_tests/steps/bundles.py:
```diff
@@ -284,3 +290,431 @@ def user_sees_updated_release_calendar_page_title_release_date_status(context: C
+            if bool(bundle_dets["add_stat_page"]) and hasattr(context, "statistical_article_page"):
+                page = context.statistical_article_page
+                BundlePage.objects.create(parent=bundle, page=page)
+                # context.bundlepages.append(BundlePageFactory(parent=bundle, page=page))
```
Commented out code left here.

AWitcherONS committed:
knownd Failing scenarios:
  functional_tests/features/bundle_ui.feature:63  A User cannot save a bundle due to duplicate schedule -- @1.1 bundles
  functional_tests/features/bundle_ui.feature:64  A User cannot save a bundle due to duplicate schedule -- @1.2 bundles
  functional_tests/features/bundle_ui.feature:172  A user can approve a bundle -- @1.1 bundles
  functional_tests/features/bundle_ui.feature:173  A user can approve a bundle -- @1.2 bundles
  functional_tests/features/bundle_ui.feature:174  A user can approve a bundle -- @1.3 bundles
  functional_tests/features/bundle_ui.feature:175  A user can approve a bundle -- @1.4 bundles
  functional_tests/features/bundle_ui.feature:208  A user cannot approve a bundle due to lack of pages or datasets -- @1.1 bundles

AWitcherONS committed:
Merge branch 'main' into feature/bundles_UI_test-CMS-333

MaciekBaron reviewed (changes requested):
I've noticed that in your latest commit you included a list of failing tests. They are failing because there are mistakes in your code. I have added a few comments to help you fix one of them ("A user can approve a bundle"). It's possible that these changes will also fix your other tests. They may also get resolved once you fix the things I mentioned in my other comments.

MaciekBaron commented on functional_tests/steps/bundles.py:
```diff
@@ -284,3 +291,431 @@ def user_sees_updated_release_calendar_page_title_release_date_status(context: C
+    context.page.get_by_text("Statistical article page").click()
+    context.page.get_by_role("button", name="Show filters").click()
+    context.page.get_by_role("button", name="Locale").click()
+    context.page.locator("#id_locale").get_by_text("English").click()
```
Just to repeat what I said a few times now: when you have a `@then` statement, you want to make some assertions. Your step says "the {role} can preview the Statistical Analysis page" _(shouldn't this say Statistical Article page anyway?)_ but all you do is you go to the page index and change a few filters.

How does that check if you can preview the page? Why are there no assertions here, to check any expectations?

`/admin/pages` is the root index of all pages. Here by default you see just a list of pages Home pages. You can check if "Statistical article page" is in the list, but the thing is, _all_ page types are listed there. You don't even check if the page you created actually appears there, you just click on the filter.

Please rethink and rewrite this step to match the requirement.

MaciekBaron commented on functional_tests/steps/bundles.py:
```diff
@@ -284,3 +291,431 @@ def user_sees_updated_release_calendar_page_title_release_date_status(context: C
+
+@then("the logged in user cannot approve a bundle due to lack of pages")
+def cannot_approve_bundle_edge_case(context: Context) -> None:
+    bundle_edit = f"Edit '{context.bundles[-1].name}'"
```
1. This test fails here as it cannot find the edit link.
2. This step should be split into at least two or three steps, e.g. "the user goes to the bundle view", "the user clicks approve", "the user is informed they cannot approve the bundle"
3. You should check that the message "Cannot approve the bundle without any pages or datasets" is visible, not just that the page could not be saved, in case something else has prevented it from being saved - be explicit in your tests.

AWitcherONS replied:
Resolved in [612ac1c](https://github.com/ONSdigital/dis-wagtail/pull/273/commits/612ac1c3d43b3d9ee6ffdea5479d80ac4d7998e8)

MaciekBaron replied:
Point 2 hasn't been addressed.

Consider your current scenario ("A user cannot approve a bundle due to lack of pages or datasets"). After your initial setup with the `Given` + `And` statements, you have essentially just two steps:

```feature
    When the <role> logs in
    Then the logged in user cannot approve a bundle due to lack of pages
 ```
 
 This is a functional/UI test. We should be explicit in what the user is trying to do. Look at existing tests in the `bundle.feature` file, like this example:
 
 https://github.com/ONSdigital/dis-wagtail/blob/48622f3e254afe5ddbd4039962a5a468bea84a34/functional_tests/features/bundle.feature#L63-L72
 
 Notice how we clearly outline all of the steps the user is taking: what they are clicking, what they are putting in etc.. Your scenario doesn't do that. Some of your other scenarios suffer from the same problem.

AWitcherONS committed:
resolve comments dec25

AWitcherONS committed:
resolve main merge jan26

AWitcherONS committed:
complete comments

AWitcherONS commented on functional_tests/features/bundle_ui.feature:
```diff
@@ -0,0 +1,208 @@
+
+Scenario Outline: A User cannot save a bundle due to duplicate schedule
+    Given there is a <role> user
+    And there is a Release Calendar page approved by <creator_role>
```
opps 
Resolved in [612ac1c](https://github.com/ONSdigital/dis-wagtail/pull/273/commits/612ac1c3d43b3d9ee6ffdea5479d80ac4d7998e8)

AWitcherONS committed:
resolve comment jan Jam 2026

AWitcherONS committed:
Merge branch 'main' into feature/bundles_UI_test-CMS-333

AWitcherONS committed:
fix lint issues Jan  2026

AWitcherONS commented on functional_tests/features/bundle_ui.feature:
```diff
@@ -0,0 +1,208 @@
+        | number_of_bundles | role                       | creator_role        | bundle_details                                                                                                                                                 |
+        | 1                 | Viewer                     | Publishing Admin    | {"role": "Viewer",             "creator_role": "Publishing Admin",  "status": "In_Review", "preview_teams": false,  "add_rel_cal": true, "add_stat_page": true}|
+        | 1                 | Publishing Officer         | Publishing Officer  | {"role": "Publishing Officer", "creator_role": "Publishing Officer", "status": "In_Review", "preview_teams": true, "add_rel_cal": true, "add_stat_page": true} |
+        | 1                 | Publishing Admin           | Publishing Admin    | {"role": "Publishing Admin",   "creator_role": "Publishing Admin",   "status": "In_Review", "preview_teams": true, "add_rel_cal": true, "add_stat_page": true} |
```
This Test has now been removed as is no longer applicable 
resolved [6c68056ac](https://github.com/ONSdigital/dis-wagtail/pull/273/changes/6c68056ac75b8d23ebc3585620d740d168c10752)

AWitcherONS commented on functional_tests/features/bundle_ui.feature:
```diff
@@ -0,0 +1,208 @@
+    And there are <number_of_bundles> bundles with <bundle_details>
+    When the <role> logs in
+    Then the logged in user can approve a bundle
+    And the bundle pages are not live
```
Yeah it didn't make sense has been remove
resolved [6c68056ac](https://github.com/ONSdigital/dis-wagtail/pull/273/changes/6c68056ac75b8d23ebc3585620d740d168c10752)

AWitcherONS committed:
resolve comments

AWitcherONS committed:
merge main

MaciekBaron commented on functional_tests/features/bundle_ui.feature:
```diff
@@ -0,0 +1,208 @@
+    Then the logged in user cannot create a bundle
+    Examples: bundles
+       | role     |
+       | Viewer   |
```
Sorry Ann, I don't understand your reply.

[Scenario Outlines](https://behave.readthedocs.io/en/latest/tutorial/#scenario-outlines) are used for parameterised scenarios, where the scenario is executed for each row in the examples table.

However, you only have one example. It's a lot more readable to do it this way:

```suggestion
Scenario: A User cannot create a bundle due to authorisation
    Given there is a Viewer user
    When the Viewer logs in
    Then the logged in user cannot create a bundle
```

It's clear what we are testing, who the user is and what they are trying to do, and we use fewer lines. It all reads better – you don't need to go to the table to check what the `<role>` parameter will be. There is no reason to "replicate" the table from the scenario outline which has multiple different roles – I see no benefit in doing that. It just makes the code clunkier and more difficult to read. If you can show me a benefit, I can definitely take it into consideration.

MaciekBaron reviewed (changes requested):
Thank you Ann for replying to some of my comments. It makes it easier to review and verify that you have addressed the issues raised. Could you please continue replying to all of the comments so that we know that everything has been addressed.

AWitcherONS committed:
merge main

AWitcherONS committed:
address comments

AWitcherONS committed:
Merge branch 'main' into feature/bundles_UI_test-CMS-333

AWitcherONS committed:
resolve comments v2

AWitcherONS committed:
resolve comments v3

AWitcherONS committed:
Move behave step  names as per comment

AWitcherONS committed:
Merge branch 'main' into feature/bundles_UI_test-CMS-333

AWitcherONS committed:
restore error scenarios

AWitcherONS committed:
fix lint

AWitcherONS committed:
Merge branch 'main' into feature/bundles_UI_test-CMS-333

AWitcherONS committed:
fix failing text in search

MaciekBaron commented on functional_tests/features/bundle_ui.feature:
```diff
@@ -0,0 +1,235 @@
+    And the logged in user can create a bundle
+    And the logged in user adds a Name to the bundle
+    When the logged in user tries to find a non existing bundle page
+    Then the logged in user gets the following message for bundle_page
```
1. What "following message"? A message doesn't follow.
2. What's `bundle_page`? The steps in your feature file should be human readable and make sense after variable substitution (this is not a variable, however). Please reconsider your approach as the sentence doesn't make sense. How about you store the search query in the context in your previous step so you can refer to it later on?

Here and on line 101

AWitcherONS replied:
Hi 
this has been replaced with bundled_page
I think its more meaningful 
resolution in [07cf10a](https://github.com/ONSdigital/dis-wagtail/pull/273/commits/07cf10afc240fd87b32f36dce6af2dd30360a77d)

AWitcherONS committed:
Merge branch 'main' into feature/bundles_UI_test-CMS-333

AWitcherONS committed:
lint issue

AWitcherONS committed:
Merge branch 'main' into feature/bundles_UI_test-CMS-333

AWitcherONS committed:
Merge branch 'main' into feature/bundles_UI_test-CMS-333

AWitcherONS committed:
Merge branch 'main' into feature/bundles_UI_test-CMS-333

AWitcherONS committed:
missed file

AWitcherONS committed:
fix add/adds

MaciekBaron commented on functional_tests/steps/bundles.py:
```diff
@@ -317,3 +327,315 @@ def user_sees_validation_error_preventing_cancellation(context: Context) -> None
+
+@then("the logged in user goes to the bundle page")
+def go_to_bundle_page(context: Context) -> None:
+    context.page.get_by_role("link", name="Bundles", exact=True).click()
```
You seem to have done a find-and-replace which affected other sentences ("can adds").

AWitcherONS committed:
fix for can see creat button

AWitcherONS committed:
revert can adds

MaciekBaron commented on functional_tests/steps/bundles.py:
```diff
@@ -359,3 +366,404 @@ def the_page_is_in_the_given_bundle_with_status(
+    expect(context.page.get_by_text("Datasets", exact=True)).to_be_visible()
+
+
+@step("the logged in user cannot approve a bundle")
```
This step is unused and it seems to be very similar to "the logged in user cannot approve a bundle due to lack of pages" on line 598.

AWitcherONS committed:
Merge branch 'main' into feature/bundles_UI_test-CMS-333

AWitcherONS committed:
Merge branch 'main' into feature/bundles_UI_test-CMS-333

AWitcherONS committed:
Merge branch 'main' into feature/bundles_UI_test-CMS-333

AWitcherONS committed:
Bundles UI resolve comments

AWitcherONS committed:
Merge branch 'main' into feature/bundles_UI_test-CMS-333

AWitcherONS committed:
remove print commented and sort

AWitcherONS committed:
improve scenario text

AWitcherONS committed:
Merge branch 'main' into feature/bundles_UI_test-CMS-333

AWitcherONS committed:
Merge branch 'main' into feature/bundles_UI_test-CMS-333

AWitcherONS committed:
add wait and resolve confusion comment

AWitcherONS committed:
Merge branch 'main' into feature/bundles_UI_test-CMS-333

AWitcherONS committed:
address git test failures

AWitcherONS committed:
address git test failures v2
